### ISSUE TITLE
Small typeInterval values cause .type() to enter truncated values

### ISSUE THREAD

**Issue by rinogo** [segmentio/nightmare#1116]
# Short description
Using the `typeInterval` setting with "small" values (e.g. 1, 10, etc.) yields inconsistent results.  The problem is that the value isn't completely entered into the field.  When an error occurs, it occurs because the end of the value is truncated (i.e. not entered completely).

# Details
Running `mocha --grep type` without any modifications to the tests consistently passes.

However, if we change the `typeInterval` that is passed to the Nightmare constructor, weird things start to happen.  To reproduce these problems, change [line 844 of test/index.js](https://github.com/segmentio/nightmare/blob/master/test/index.js#L844) to be:

```
      nightmare = Nightmare({typeInterval: 10});
```

After making the aforementioned change, execute just the tests that mention "type":
```
mocha --grep type
```

On my system, the "should type integer" (and only that test) test fails *nearly* every time.  **Sometimes, it passes.**  I think in about 60 executions, it has failed 55 times (only passed five times).  Interestingly enough, three of the times that the tests passed were all in a row, and were all immediately after Cmd-Tabbing back to Terminal.  This may suggest that the erratic behavior is correlated somehow with system load.

An example of a failed test run:

```
MachineName:nightmare rinogo$ mocha --grep type


  Nightmare
    manipulation
      ✓ should not inject unsupported types onto the page
      ✓ should type
      1) should type integer
      ✓ should type object
      ✓ should not type in a nonexistent selector
      ✓ should type and click (1540ms)
      ✓ should type and click several times (2370ms)


  6 passing (8s)
  1 failing

  1) Nightmare manipulation should type integer:

      AssertionError: expected '1' to equal '10'
      + expected - actual

      -1
      +10
      
      at Context.<anonymous> (test/index.js:942:22)
      at next (native)
      at onFulfilled (node_modules/mocha-generators/node_modules/co/index.js:65:19)
```

Now, if we use `typeInterval: 1` instead, we get different results.  With this setting, "should type integer" consistently passes.  However, "should type" usually fails and "should type object" always fails (or, more correctly, I've never personally seen it pass with `typeInterval: 1`).  Example:

```
MachineName:nightmare rinogo$ mocha --grep type


  Nightmare
    manipulation
      ✓ should not inject unsupported types onto the page
      1) should type
      ✓ should type integer
      2) should type object
      ✓ should not type in a nonexistent selector
      ✓ should type and click
      ✓ should type and click several times (2302ms)


  5 passing (8s)
  2 failing

  1) Nightmare manipulation should type:

      AssertionError: expected 'nightmar' to equal 'nightmare'
      + expected - actual

      -nightmar
      +nightmare
      
      at Context.<anonymous> (test/index.js:924:20)
      at next (native)
      at onFulfilled (node_modules/mocha-generators/node_modules/co/index.js:65:19)

  2) Nightmare manipulation should type object:

      AssertionError: expected '[object O' to equal '[object Object]'
      + expected - actual

      -[object O
      +[object Object]
      
      at Context.<anonymous> (test/index.js:963:22)
      at next (native)
      at onFulfilled (node_modules/mocha-generators/node_modules/co/index.js:65:19)
```

Note the specific error for the "should type" test: `AssertionError: expected 'nightmar' to equal 'nightmare'`.  In other words, the text wasn't completely entered.  This is the behavior that led me to investigate this problem in the first place.  Our app was yielding inconsistent results, and when running with `show: true`, I noticed certain strings not being entered completely.  These test results corroborate that observation.

I imagine using different values for `typeInterval` will yield distinct results, failing specific tests at specific intervals.

Thoughts?

**Comment by rinogo** [comment]
By the way, in anticipation of the question regarding why this matters:

The default `typeInterval` value is 100 ms.  If a test/app/script only types 10 characters, then a full second of its execution time is spent simply waiting for characters to be typed.  If it types 100 characters, then ten seconds is spent waiting.

Being able to use `0` for `typeInterval` would be ideal (currently not allowed), but `1` is acceptable.  Even `10` or something close to it may be acceptable.

**Comment by rinogo** [comment]
Addendum: I also seem to experience these problems (even with the default typeInterval!) when trying to run multiple simultaneous processes of node/NightmareJS (via `xvfb-run`).  Yet another data point that suggests that this is somehow related to load.

**Comment by rinogo** [comment]
I've had a bit more time to look into this.

The `.type()` function calls `done()` as soon as it detects that the `chars` buffer is empty ([Source](https://github.com/segmentio/nightmare/blob/master/lib/runner.js#L375)).  I believe what's happening is that `type()` is getting called again (by the [recursive call](https://github.com/segmentio/nightmare/blob/master/lib/runner.js#L401)) before the events have propagated fully, causing the characters to actually be typed by Electron.

I believe an acceptable fix to this would be to add a check before calling `done()` that the characters that have been successfully typed match what we intended to type (e.g. [the original `value` parameter](https://github.com/segmentio/nightmare/blob/master/lib/runner.js#L369)).  If they match, `done()` can be called.  If they don't match, `setTimeout()` should be called again.

I don't know much about Electron, so I'm not sure how to implement this.  Perhaps someone with a bit more knowledge can do so.

**Regardless, I've chosen to use `.insert()` instead of `.type()`.  In looking into this issue, I stumbled across the `.insert()` function which is exactly what I'm looking for - Essentially, `.type()` with a `typeInterval` of `0`.**  Given that [`.insert()` simply inserts the text and calls `done()`](https://github.com/segmentio/nightmare/blob/master/lib/runner.js#L409), it doesn't exhibit the same problems as `.type()`.
